vyper/semantics/validation/local.py

```diff
@@ -162,47 +162,51 @@ class FunctionNodeVisitor(VyperNodeVisitorBase):
     scope_name = "function"
 
     def __init__(
-        self, vyper_module: vy_ast.Module, fn_node: vy_ast.FunctionDef, namespace: dict
+        self,
+        vyper_module: vy_ast.Module,
+        namespace: dict,
+        fn_node: vy_ast.FunctionDef = None,
     ) -> None:
         self.vyper_module = vyper_module
-        self.fn_node = fn_node
+        self.fn_node = fn_node if fn_node else None
         self.namespace = namespace
-        self.func = fn_node._metadata["type"]
-        self.annotation_visitor = StatementAnnotationVisitor(fn_node, namespace)
+        self.func = fn_node._metadata["type"] if fn_node else None
+        self.annotation_visitor = StatementAnnotationVisitor(namespace, fn_node=fn_node)
         self.expr_visitor = _LocalExpressionVisitor()
-        namespace.update(self.func.arguments)
-
-        if self.func.mutability == StateMutability.PURE:
-            node_list = fn_node.get_descendants(
-                vy_ast.Attribute,
-                {
-                    "value.id": set(CONSTANT_ENVIRONMENT_VARS.keys()).union(
-                        set(MUTABLE_ENVIRONMENT_VARS.keys())
-                    )
-                },
-            )
-            if node_list:
-                raise StateAccessViolation(
-                    "not allowed to query contract or environment variables in pure functions",
-                    node_list[0],
+        if fn_node:
```

Comment:
this increases nesting. in general it's better to fastpath out if possible. e.g.
```
if fn_node is None:
    return
...
```